frontend/src/metabase/lib/expressions/typechecker.js

```diff
@@ -99,3 +99,108 @@ export function typeCheck(cst, rootType) {
   checker.visit(cst);
   return { typeErrors: checker.errors };
 }
+
+/*
+
+  Create a copy of the syntax tree where the unnecessary intermediate nodes
+  are not present anymore.
+
+  Example:
+  For a simple expression "42", the syntax tree produced by the parser is
+
+  expression <-- this is the root node
+    relationalExpression
+      additionExpression
+        multiplicationExpression
+          atomicExpression
+            numberLiteral
+
+  Meanwhile, the compact variant of the syntax tree:
+
+    numberLiteral
+
+*/
+
+export function compactSyntaxTree(node) {
+  if (!node) {
+    return;
+  }
+  const { name, children } = node;
+  let result = { name, children };
+
+  switch (name) {
+    case "any":
+    case "aggregation":
+    case "atomicExpression":
+    case "boolean":
+    case "booleanExpression":
+    case "string":
+      {
+        if (children.expression) {
+          const expression = children.expression.map(compactSyntaxTree);
+          result =
```

Comment:
Unimportant detail, but could these `result =` assignments be replaced with `return`s?

Reply:
Definitely. I'll change it accordingly.